lib_eio/flow.ml

```diff
@@ -60,8 +60,11 @@ let string_source s = cstruct_source [Cstruct.of_string s]
 class virtual sink = object (_ : <Generic.t; ..>)
   method probe _ = None
   method virtual copy : 'a. (#source as 'a) -> unit
+  method virtual write_from : Cstruct.t -> unit
 end
 
+let write (t : #sink) (buf : Cstruct.t) = t#write_from buf
```

Comment:
With the change above, we can provide both `write_all` and `write_single` here, like in `File`.

Reply:
My first version had both, ha, I'll add it back.

Reply:
The problem is libuv doesn't give us a way to do `write_single`, it will queue the next write and attempt to write all of them.
```
      int uv_write(uv_write_t *req, uv_stream_t *handle, const uv_buf_t bufs[], unsigned int nbufs, uv_write_cb cb)
              Write data to stream. Buffers are written in order. Example:

                 void cb(uv_write_t* req, int status) {
                     /* Logic which handles the write result */
                 }

                 uv_buf_t a[] = {
                     { .base = "1", .len = 1 },
                     { .base = "2", .len = 1 }
                 };

                 uv_buf_t b[] = {
                     { .base = "3", .len = 1 },
                     { .base = "4", .len = 1 }
                 };

                 uv_write_t req1;
                 uv_write_t req2;

                 /* writes "1234" */
                 uv_write(&req1, stream, a, 2, cb);
                 uv_write(&req2, stream, b, 2, cb);

              NOTE:
                 The memory pointed to by the buffers must remain valid until the callback gets called.  This also holds for uv_write2().
```

And no, it doesn't return bytes written, it's just 0 or error :/

Reply:
So the option is to either lie (write_single = write_all) in libuv, and god knows how many writes were done, or not provide a write_single.

Reply:
should `copy` have a default implementation in terms of `write_from`?

Reply:
libuv is callback based. I think the `return 0` just means the write operation was scheduled. The callback gets the number of bytes written.

https://github.com/aantron/luv/blob/master/src/stream.mli says:

> The second argument passed to the callback is the number of bytes
> written. libuv has an internal queue of writes, in part to implement retry.
> This means that writes can be partial at the libuv (and Luv) API level, so
> it is possible to receive both an [Error] result, and for some data to have
> been successfully written.

However, for files it looks like `write` behaves normally (either returns an error or a byte count, but not both).

So, yeah, maybe we shouldn't provide `write_single`. But then we don't need a write method at all - `Flow.write flow buf` can just be `Flow.copy (cstruct_source [buf]) flow`.

Possibly we just need to add `exception Write_error of int * exn` to report how many bytes got written, in case anyone cares.

Reply:
> libuv is callback based. I think the `return 0` just means the write operation was scheduled. The callback gets the number of bytes written.
> 

Even though it's callback based, the code tries to do a write before setting up an event, so we can reason about that only:
`uv_write()->uv_write2()->empty_queue->uv__write()->uv__try_write()->uv__writev()->writev(2)`

> https://github.com/aantron/luv/blob/master/src/stream.mli says:
> 
> > The second argument passed to the callback is the number of bytes
> > written. libuv has an internal queue of writes, in part to implement retry.
> > This means that writes can be partial at the libuv (and Luv) API level, so
> > it is possible to receive both an [Error] result, and for some data to have
> > been successfully written.

This is a bit bad worded, what he means is that you can get an error _and_ you also get back how many bytes you successfully wrote before from previous writes, but the callback is only called once. There is no way to distinguish on success if the requested amount occurred in one or N calls.

With this I stand by my previous assessment (humbly), luv does not give us an API to do something in one `write` or give us control back in-between writes. In a sense they really treat a stream as a stream.

In fact, in order to retrieve the amount of partial bytes on write, he must be reaching into what libuv calls "private", cause the callback for write is not like read:
```
     typedef void (*uv_write_cb)(uv_write_t *req, int status)
              Callback called after data was written on a stream. status will be 0 in case of success, < 0 otherwise.
       typedef void (*uv_read_cb)(uv_stream_t *stream, ssize_t nread, const uv_buf_t *buf)
              Callback called when data was read on a stream.
```
http://docs.libuv.org/en/v1.x/stream.html#c.uv_write_cb

I can dig up luv later but he is probably getting from:
```
struct uv_write_s { /* haesbaert: this is write_cb_t without the useless typedef */
  UV_REQ_FIELDS
  uv_write_cb cb;
  uv_stream_t* send_handle; /* TODO: make private and unix-only in v2.x. */
  uv_stream_t* handle;
  UV_WRITE_PRIVATE_FIELDS
};
#define UV_WRITE_PRIVATE_FIELDS   
  void* queue[2]; 
  unsigned int write_index; /* haesbaert: this guy is likely the second argument to stream */
  uv_buf_t* bufs;
  unsigned int nbufs; 
  int error;
  uv_buf_t bufsml[4];

> 
> However, for files it looks like `write` behaves normally (either returns an error or a byte count, but not both).
> 
> So, yeah, maybe we shouldn't provide `write_single`. But then we don't need a write method at all - `Flow.write flow buf` can just be `Flow.copy (cstruct_source [buf]) flow`.
> 
> Possibly we just need to add `exception Write_error of int * exn` to report how many bytes got written, in case anyone cares.

I'd be against that, I like the idea of `write_all` + `write_single`, one means "I don't care how" and the other means "I need it to be in one".

Reply:
Leme rephrase that, I'm fine with not having single + all, but having the Write_error is not necessary imho.

Reply:
Yeah, no need for `Write_error` (or maybe `Partial_write`?) in this PR. I've never needed it, so we'll wait and see if anyone else does.

Note: `fallback_copy` won't be used if the source is a `cstruct_source` (it'll use `copy_with_rsb` instead, which behaves the way you want already). Though the luv backend doesn't do this optimisation yet.

I think it's fine to add a `write` method as an optimisation for this (would be interesting to know what the speed difference is).